scene/3d/collision_shape.cpp

```diff
@@ -199,23 +209,32 @@ CollisionShape::~CollisionShape() {
 	//VisualServer::get_singleton()->free(indicator);
 }
 
-void CollisionShape::_create_debug_shape() {
+void CollisionShape::_update_debug_shape() {
+	if (debug_shape_dirty) {
+		debug_shape_dirty = false;
+		if (debug_shape) {
+			debug_shape->queue_delete();
+			debug_shape = NULL;
+		}
 
-	if (debug_shape) {
-		debug_shape->queue_delete();
-		debug_shape = NULL;
-	}
+		Ref<Shape> s = get_shape();
 
-	Ref<Shape> s = get_shape();
+		if (s.is_null())
+			return;
 
-	if (s.is_null())
-		return;
+		Ref<Mesh> mesh = s->get_debug_mesh();
 
-	Ref<Mesh> mesh = s->get_debug_mesh();
+		MeshInstance *mi = memnew(MeshInstance);
+		mi->set_mesh(mesh);
 
-	MeshInstance *mi = memnew(MeshInstance);
-	mi->set_mesh(mesh);
+		add_child(mi);
+		debug_shape = mi;
+	}
+}
 
-	add_child(mi);
-	debug_shape = mi;
+void CollisionShape::_shape_changed() {
+	if (get_tree()->is_debugging_collisions_hint()) {
+		debug_shape_dirty = true;
```

Comment:
This does not make much sense and its going to create a ton of call_deferreds. You should exit this function and not do call deferred again if debug_shape_dirty is already true

Reply:
how should i do to check the dirty param at each frame ? I try to receive NOTIFICATION_PROCESS but it seems not work.